### PR TITLE
names, concepts, and more

### PR DESCRIPTION
**Pull request by NeilMenne**
For code organization and optimization, I'm sure you can help me out. I haven't looked at the tests yet either, so I probably destroyed those. I'll circle back to testing tomorrow or Monday. Instead, I did the following:

1. Renamed now, update. This was mostly because I was working directly against the paper on my version and just brought a lot of it over wholesale (now and update are fine with me if you prefer).
2. Lifted node_id as a parameter of interest in construction, `Timestamp.new/0`, and in messaging, `HLClock.recv_timestamp/2`, and in comparison.
3. According to the paper, HLCs are bounded on both sides by ϵ. This means that it is entirely possible that your sends could also fail. For example, you could see a timestamp in the "near future," and your physical time doesn't catch up because NTP is not running. Drift could accumulate in this case. This should probably be the kind of failure that propagates up, but we'll worry about that when we get to `GenServer` and whatnot.
4. Renamed logical to counter. This was mostly easier to read relative to the paper as `l` denotes the logical clock which is separate from the counter, `c`. Again, I'm not 100% attached to it; it was just easier to work against the paper.

*NOTE:* compare could probably be simplified given `Timestamp.encode/1`. As the relative ordering of time, counter, and node_id is baked into how a timestamp is packed. I don't know about the performance though. Thoughts?

### CODE REVIEW

**Comment by NeilMenne** [comment]
After stepping away, I could also be convinced that passing the clock function isn't necessary for new which would get rid of the need for the struct altogether.

**Comment by keathley** [comment]
I like `send_timestamp` and `recv_timestamp` more then `now` and `update`. I used those names since `send` and `receive` are already things in erlang/elixir.

Good point about accounting for drift on the "send" side. I like `counter` over `logical`.

As far as performance goes, pattern matching is pretty much always going to win. But for ease of understanding I'm fine with using encode to do comparisons. If it really is too slow then we can optimize it at that point.

I'll go through the code and leave some notes. Overall it looks good. There are some elixir style things that I can point out but aren't really blockers for merging this in IMO.

File: lib/hlclock.ex [hunk 1]
```diff
@@ -1,50 +1,104 @@
...
-      Timestamp.increment(clock)
-    else
-      Timestamp.new(pt, 0)
+  @doc "generate a new timestamp"
+  def send_timestamp(%Clock{curr_ts: old, clock_fn: cfn} = clock) do
+    pt = cfn.()
+    nt = max(old.time, pt)
+    nc = advance(old, nt)
+
+    cond do
```
Comment (keathley) [inline lib/hlclock.ex]: I'd probably turn this whole block into a `with`:

```elixir
with {:ok, clock}   <- handle_drift(clock, nt, pt),
      {:ok, new_ts} <- Timestamp.new(nt, nc, old.node_id),
      new_clock     <- put_in(clock, [:curr_ts], new_ts) do
  {:ok, new_clock}
end
```

File: lib/hlclock.ex [hunk 2]
```diff
@@ -1,50 +1,104 @@
...
+        old.counter + 1
+      true ->
+        0
+    end
   end
 
-  def update(local, foreign, pt \\ default_clock()) do
-    max_pt = max(pt, max(local.time, foreign.time))
+  @doc "receive a remote timestamp and merge with local"
+  def recv_timestamp(%Clock{clock_fn: cfn, curr_ts: local} = clock, msg) do
```
Comment (keathley) [inline lib/hlclock.ex]: Is msg here a `Timestamp`?

File: lib/hlclock.ex [hunk 3]
```diff
@@ -1,50 +1,104 @@
...
-        max_logical = max(local.logical, foreign.logical) + 1
-        Timestamp.new(max_pt, max_logical)
-      foreign.time > local.time && remote_drifted?(foreign, pt) ->
-        local
-      max_pt == local.time ->
-        Timestamp.new(max_pt, local.logical+1)
-      max_pt == foreign.time ->
-        Timestamp.new(max_pt, foreign.logical+1)
+      local.node_id == msg.node_id ->
+        {:error, :duplicate_node_id}
```
Comment (keathley) [inline lib/hlclock.ex]: I like that we're actually handling all of these errors now.

File: lib/hlclock.ex [hunk 4]
```diff
@@ -1,50 +1,104 @@
...
+      drift?(msg.time, pt) ->
+        {:error, :remote_drift_violation}
+      drift?(max_pt, pt) ->
+        {:error, :clock_drift_violation}
       true ->
-        Timestamp.new(max_pt, 0)
+        new_log = merge_logical(max_pt, local, msg)
```
Comment (keathley) [inline lib/hlclock.ex]: Similar to the above comment I'd probably do this as part of the `with` statement.

Comment (NeilMenne) [inline lib/hlclock.ex]: The multi-with is nice. I'll incorporate that where appropriate

File: lib/hlclock.ex (hunk 2 above)
Comment (NeilMenne) [inline lib/hlclock.ex]: I think I'm going to drop the struct. `msg` is a timestamp, but I'll go with something like remote vs local.

File: lib/hlclock.ex (hunk 1 above)
Comment (NeilMenne) [inline lib/hlclock.ex]: that's concise

**Comment by NeilMenne** [comment]
The logic that is associated with the struct would be better suited to the process level. I will make it so!

File: lib/hlclock.ex (hunk 4 above)
Comment (keathley) [inline lib/hlclock.ex]: Yeah `with` is my favorite thing they've added to the language. If your functions return meaningful `{:error, error.t}` tuples then you can just let the errors fall through. If you want to wrap those errors into something more meaningful (for instance if you're dealing with a dependency and don't wanna return their error) you can do something like this:

```elixir
with {:ok, clock} <- handle_drift(clock),
        {:ok, result} <- do_some_thing(clock) do
else
  {:error, :some_error} ->
    {:error, :failed_to_build_clock}
end
```

**Comment by NeilMenne** [comment]
@keathley, finally unborked the tests and whatnot

**Pull request merged**